File: LeetCode/Java/Medium/22. Generate Parentheses/Solution.java
import java.util.*;

class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        backtrack(result, "", 0, 0, n);
        return result;
    }

    private void backtrack(List<String> result, String current, int open, int close, int n) {
        // base case
        if (current.length() == 2 * n) {
            result.add(current);
            return;
        }

        // add opening bracket if possible
        if (open < n) {
            backtrack(result, current + "(", open + 1, close, n);
        }

        // add closing bracket if valid
        if (close < open) {
            backtrack(result, current + ")", open, close + 1, n);
        }
    }
}

File: LeetCode/C++/Medium/22. Generate Parentheses/Solution.cpp
class Solution {
public:
    void fun(vector<string> &ans, string s,int open,int close,int n){
        if(open == n && close == n){
           ans.push_back(s);
           return;
        }
        if(open < n){
            fun(ans,s + '(',open + 1,close,n);
        }
        if(close < open){
            fun(ans,s + ')',open, close + 1,n);
        }

    }
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        fun(ans,"",0,0,n);
        return ans;
        
    }
};